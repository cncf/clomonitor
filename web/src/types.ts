export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
