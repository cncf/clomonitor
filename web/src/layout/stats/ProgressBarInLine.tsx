import { ReportOption } from '../../types';
import getCategoryColor from '../../utils/getCategoryColor';
import styles from './ProgressBarInLine.module.css';

interface Props {
  title: string;
  icon: JSX.Element;
  value: number;
  name: ReportOption;
  onSelectCheck: (name: ReportOption) => void;
}

const ProgressBarInLine = (props: Props) => {
  const color = getCategoryColor(props.value);

  return (
    <div className={`d-flex flex-column ${styles.wrapper}`}>
      <div className={`d-flex flex-row align-items-center ${styles.progressWrapper}`}>
        <div className={`me-1 me-md-2 position-relative ${styles.icon}`}>{props.icon}</div>
        <div className={styles.progressTitle}>
          <span className="d-inline-block d-md-none">{props.title}</span>
          <span className="d-none d-md-inline-block">
            <button onClick={() => props.onSelectCheck(props.name)} className={`btn btn-link p-0 ${styles.btn}`}>
              {props.title}
            </button>
          </span>
        </div>
        <div className="flex-grow-1 ms-2">
          <div className={`progress rounded-0 ${styles.progress}`}>
            <div
              className="progress-bar"
              role="progressbar"
              style={{ width: `${props.value || 1}%`, backgroundColor: `var(--rm-${color})` }}
            />
          </div>
        </div>
        <div className={`ps-2 lh-1 text-end fw-bold lightText ${styles.scoreWrapper}`}>{props.value}%</div>
      </div>
    </div>
  );
};

export default ProgressBarInLine;
