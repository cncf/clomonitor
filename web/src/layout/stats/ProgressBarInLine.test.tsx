import { render, screen } from '@testing-library/react';

import { ReportOption } from '../../types';
import ProgressBarInLine from './ProgressBarInLine';

const defaultProps = {
  title: 'Governance',
  icon: <>icon</>,
  value: 75,
  name: ReportOption.Governance,
  onSelectCheck: jest.fn(),
};

describe('ProgressBarInLine', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<ProgressBarInLine {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders bar', () => {
      render(<ProgressBarInLine {...defaultProps} />);

      expect(screen.getAllByText('Governance')).toHaveLength(2);
      expect(screen.getByText('icon')).toBeInTheDocument();
      expect(screen.getByText('75%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-green)' });
      expect(progressBar).toHaveStyle({ width: '75%' });
    });

    it('renders bar - 20', () => {
      render(<ProgressBarInLine {...defaultProps} value={20} />);

      expect(screen.getByText('20%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-red)' });
      expect(progressBar).toHaveStyle({ width: '20%' });
    });

    it('renders bar - 40', () => {
      render(<ProgressBarInLine {...defaultProps} value={40} />);

      expect(screen.getByText('40%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-orange)' });
      expect(progressBar).toHaveStyle({ width: '40%' });
    });

    it('renders bar - 60', () => {
      render(<ProgressBarInLine {...defaultProps} value={60} />);

      expect(screen.getByText('60%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-yellow)' });
      expect(progressBar).toHaveStyle({ width: '60%' });
    });

    it('renders bar - 80', () => {
      render(<ProgressBarInLine {...defaultProps} value={80} />);

      expect(screen.getByText('80%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-green)' });
      expect(progressBar).toHaveStyle({ width: '80%' });
    });
  });
});
