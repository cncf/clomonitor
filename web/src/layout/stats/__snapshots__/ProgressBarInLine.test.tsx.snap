// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarInLine creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column wrapper"
  >
    <div
      class="d-flex flex-row align-items-center progressWrapper"
    >
      <div
        class="me-1 me-md-2 position-relative icon"
      >
        icon
      </div>
      <div
        class="progressTitle"
      >
        <span
          class="d-inline-block d-md-none"
        >
          Governance
        </span>
        <span
          class="d-none d-md-inline-block"
        >
          <button
            class="btn btn-link p-0 btn"
          >
            Governance
          </button>
        </span>
      </div>
      <div
        class="flex-grow-1 ms-2"
      >
        <div
          class="progress rounded-0 progress"
        >
          <div
            class="progress-bar"
            role="progressbar"
            style="width: 75%;"
          />
        </div>
      </div>
      <div
        class="ps-2 lh-1 text-end fw-bold lightText scoreWrapper"
      >
        75%
      </div>
    </div>
  </div>
</DocumentFragment>
`;
